README.md
# Zway-WeatherUnderground

Zway Automation module for fetching current condition and forecasts from the
WeatherUnderground API. Creates virtual devices for humidity, wind, current
conditions, UV index, barometric pressure and forecasts.

See http://www.wunderground.com/weather/api/d/docs for details.

# Configuration

## location

Accepts any format that constitutes a valid API query.

*   AIRPORTCODE (eg. "VIE")
*   COUNTRY/CITY (eg. "Austria/Vienna")
*   pws:PWSID (personal weather station ID, eg. "pws:IWIEN335")
*   "autoip" (based on geo ip)
*   LAT,LNG (eg. "48.214,16.401")

## unitTemperature

Display temperatures in Celsius or Fahrenheit

## unitSystem

Display metric or imperial units for air pressure and wind speed

## apiKey

Required API key for accessing the service. See
http://www.wunderground.com/weather/api/ for obtaining an API key.

## uvDevice, windDevice, humidityDevice, barometerDevice

Flag that sets if devices should be created

# Virtual Devices

This module creates up to six virtual devices

## Current conditions

Displays the current condition as an icon and the current temperature.
Additionally the following metrics are set

*    metrics:level Current temperature
*    metrics:temperature
*    metrics:condition: Current condition
*    metrics:conditiongroup: fair,neutral,poor or snow
*    metrics:feelslike
*    metrics:weather: Current weather summary
*    metrics:pop: probability of precipitation
*    metrics:high: expected high temperature today
*    metrics:low: expected low temperature today
*    metrics:percipintensity: Rain intensity in mm or inches / 1h
*    metrics:raw: raw current conditions data returned by the API

## Forecast

Displays the forecasted condition as an icon, and the expected temperature
range.

*    metrics:level Forecast temperature range
*    metrics:condition: Current condition
*    metrics:conditiongroup: fair,neutral,poor or snow
*    metrics:weather: Current weather summary
*    metrics:pop: probability of precipitation
*    metrics:high: expected high temperature today
*    metrics:low: expected low temperature today
*    metrics:raw: raw forecast data returned by the API

## Wind

Displays the current wind speed. Wind strength is indicated by the icon.

*    metrics:level: Wind speed (wind / 2 + windgust / 2)
*    metrics:dir: Wind direction
*    metrics:wind: Base wind speed
*    metrics:wind_avg: Wind average of last three updates
*    metrics:windgust: Wind gust speeds
*    metrics:winddregrees: Wind degrees
*    metrics:beaufort: Wind strength in beaufort [0-12]

## Humidity

Displays the current humidity.

## UV

Displays the current UV index [0-11+]

*    metrics:uv_avg: UV index average of last three updates
*    metrics:solarradiation: in watts/m2
*    metrics:solarradiation_avg: Radiation average of last three updates

## Barometer

Displays the current air pressure. Rising and falling pressure is also
indicated by the icon.

*    metrics:trend: Pressure trend

# Events

No events are emitted

# Installation

The prefered way of installing this module is via the "Zwave.me App Store"
available in 2.2.0 and higher. For stable module releases no access token is
required. If you want to test the latest pre-releases use 'k1_beta' as
app store access token.

For developers and users of older Zway versions installation via git is
recommended.

```shell
cd /opt/z-way-server/automation/userModules
git clone https://github.com/maros/Zway-WeatherUnderground.git WeatherUnderground --branch latest
```

To update or install a specific version
```shell
cd /opt/z-way-server/automation/userModules/WeatherUnderground
git fetch --tags
# For latest released version
git checkout tags/latest
# For a specific version
git checkout tags/1.02
# For development version
git checkout -b master --track origin/master
```

# License

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or any
later version.

Parts of the barometer icon are from (http://www.flaticon.com/authors/yannick)
, licensed under the Creative Commons 3.0 license.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
GNU General Public License for more details.


CHANGES
1.13 2019 Apr 26, Fri
    * Add deprecation notice
    * Improved location handling
    * Fetch weather retry

1.12 2017 Jun 24, Sat
    * Handle sunset/rise only when data is available
    * Update copyright
    * Improved wind calculations

1.11 2016 May 31, Tue
    * Update labels
    * Cleanup unit
    * Fix average calculation
    * Fix solarradiation device
    * Fix defaults
    * Improve texts
    * Update module title
    * Update probeType

1.1 2016 Apr 21, Thu
    * Temperature falling or rising
    * Minor fixups
    * Delay initial weather fetch to speed up startup
    * Handle manual weather updates
    * Set probeType
    * Clarify config, fix typos
    * Clarify location config option
    * Update installation howto

1.09 2016 Feb 24, Wed
    * Load data on first init
    * Remove unused default
    * Fix HTTP error reporting
    * Add doc link

1.08 2016 Jan 18, Mon
    * Add forecast low/high temp devices
    * Rename title/description key to make module work in appstore
    * Update probeTitle
    * Update readme & homepage

1.07 2015 Dec 20, Sun
    * Add metrics:percipintensity
    * Update title

1.06 2015 Dec 19, Sat
    * Improve handling of wind
    * cleanup jshint issues
    * German translation

1.05 2015 Nov 27, Fri
    * Improve wind calculation
    * Update wind -> beaufort
    * Replace logo to honour license

1.04 2015 Nov 15, Sun
    * Set wind level in beaufort
    * Update documentation
    * Use standard probe titles
    * Unify variable names (to match with ForecastIO)

1.03 2015 Nov 03, Tue
    * Fix detection of last query
    * Fix module schema
    * Add barometer
    * Cleanup device config flags

1.02 2015 Nov 02, Mon
    * Fix device creation
    * Fix initial init
    * Fix UV setpoints
	* UV device
	* Specify which devices to create (wind, uv, humidity)
    * Store values as numbers, not strings
    * Change wind probeTitle
    * Load lang file only once

1.01 2015 Oct 25, Sun
    * Don't treat fog as rain
	* Don't query on startup if last query is recent
	* Raw forecast data

1.00 2015 Oct 16, Fri
    * Initial release


underscore_deep_extend.js
/*  Copyright (C) 2012-2014  Kurt Milam - http://xioup.com | Source: https://gist.github.com/1868955
 *   
 *  Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the "Software"), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:
 *  The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.
 *
 *  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
**/

// Based conceptually on the _.extend() function in underscore.js ( see http://documentcloud.github.com/underscore/#extend for more details )

function deepExtend(obj) {
  var parentRE = /#{\s*?_\s*?}/,
  slice = Array.prototype.slice;

  _.each(slice.call(arguments, 1), function(source) {
    for (var prop in source) {
      if (_.isUndefined(obj[prop]) || _.isFunction(obj[prop]) || _.isNull(source[prop]) || _.isDate(source[prop])) {
        obj[prop] = source[prop];
      }
      else if (_.isString(source[prop]) && parentRE.test(source[prop])) {
        if (_.isString(obj[prop])) {
          obj[prop] = source[prop].replace(parentRE, obj[prop]);
        }
      }
      else if (_.isArray(obj[prop]) || _.isArray(source[prop])){
        if (!_.isArray(obj[prop]) || !_.isArray(source[prop])){
          throw new Error('Trying to combine an array with a non-array (' + prop + ')');
        } else {
          obj[prop] = _.reject(_.deepExtend(_.clone(obj[prop]), source[prop]), function (item) { return _.isNull(item);});
        }
      }
      else if (_.isObject(obj[prop]) || _.isObject(source[prop])){
        if (!_.isObject(obj[prop]) || !_.isObject(source[prop])){
          throw new Error('Trying to combine an object with a non-object (' + prop + ')');
        } else {
          obj[prop] = _.deepExtend(_.clone(obj[prop]), source[prop]);
        }
      } else {
        obj[prop] = source[prop];
      }
    }
  });
  return obj;
};

_.mixin({ 'deepExtend': deepExtend });

/**
 * Dependency: underscore.js ( http://documentcloud.github.com/underscore/ )
 *
 * Mix it in with underscore.js:
 * _.mixin({deepExtend: deepExtend});
 * 
 * Call it like this:
 * var myObj = _.deepExtend(grandparent, child, grandchild, greatgrandchild)
 *
 * Notes:
 * Keep it DRY.
 * This function is especially useful if you're working with JSON config documents. It allows you to create a default
 * config document with the most common settings, then override those settings for specific cases. It accepts any
 * number of objects as arguments, giving you fine-grained control over your config document hierarchy.
 *
 * Special Features and Considerations:
 * - parentRE allows you to concatenate strings. example:
 *   var obj = _.deepExtend({url: "www.example.com"}, {url: "http://#{_}/path/to/file.html"});
 *   console.log(obj.url);
 *   output: "http://www.example.com/path/to/file.html"
 *
 * - parentRE also acts as a placeholder, which can be useful when you need to change one value in an array, while
 *   leaving the others untouched. example:
 *   var arr = _.deepExtend([100,    {id: 1234}, true,  "foo",  [250, 500]],
 *                          ["#{_}", "#{_}",     false, "#{_}", "#{_}"]);
 *   console.log(arr);
 *   output: [100, {id: 1234}, false, "foo", [250, 500]]
 *
 * - The previous example can also be written like this:
 *   var arr = _.deepExtend([100,    {id:1234},   true,  "foo",  [250, 500]],
 *                          ["#{_}", {},          false, "#{_}", []]);
 *   console.log(arr);
 *   output: [100, {id: 1234}, false, "foo", [250, 500]]
 *
 * - And also like this:
 *   var arr = _.deepExtend([100,    {id:1234},   true,  "foo",  [250, 500]],
 *                          ["#{_}", {},          false]);
 *   console.log(arr);
 *   output: [100, {id: 1234}, false, "foo", [250, 500]]
 *
 * - Array order is important. example:
 *   var arr = _.deepExtend([1, 2, 3, 4], [1, 4, 3, 2]);
 *   console.log(arr);
 *   output: [1, 4, 3, 2]
 *
 * - You can remove an array element set in a parent object by setting the same index value to null in a child object.
 *   example:
 *   var obj = _.deepExtend({arr: [1, 2, 3, 4]}, {arr: ["#{_}", null]});
 *   console.log(obj.arr);
 *   output: [1, 3, 4]
 *
 **/

index.js
/*** WeatherUnderground Z-Way HA module *******************************************

Version: 1.13
(c) Maroš Kollár, 2015-2017
-----------------------------------------------------------------------------
Author: Maroš Kollár <maros@k-1.com>
Description:
    This module checks weather updates via weatherundergound.com

******************************************************************************/

/*jshint loopfunc: true */

function WeatherUnderground (id, controller) {
    // Call superconstructor first (AutomationModule)
    WeatherUnderground.super_.call(this, id, controller);

    this.location           = undefined;
    this.apiKey             = undefined;
    this.unitTemperature    = undefined;
    this.unitSystem         = undefined;
    this.timer              = undefined;
    this.update             = undefined;
    this.devices            = {};
}

inherits(WeatherUnderground, AutomationModule);

_module = WeatherUnderground;

// ----------------------------------------------------------------------------
// --- Module instance initialized
// ----------------------------------------------------------------------------

WeatherUnderground.prototype.deviceTypes = ['wind','uv','humidity','barometer','forecastLow','forecastHigh'];
WeatherUnderground.prototype.windBeaufort = [
    1.1,    // 0
    5.5,
    11.9,
    19.7,
    28.7,
    38.8,
    49.9,   // 6
    61.8,
    74.6,
    88.1,
    102.4,
    117.4,
    Number.POSITIVE_INFINITY
];
WeatherUnderground.prototype.windIcons = [
    1,
    3,
    6,
    Number.POSITIVE_INFINITY
];
WeatherUnderground.prototype.init = function (config) {
    WeatherUnderground.super_.prototype.init.call(this, config);

    var self = this;

    self.location           = config.location.toString();
    self.apiKey             = config.apiKey.toString();
    self.unitTemperature    = config.unitTemperature.toString();
    self.unitSystem         = config.unitSystem.toString();
    self.langFile           = self.controller.loadModuleLang("WeatherUnderground");
    var scaleTemperature    = self.unitTemperature === "celsius" ? '°C' : '°F';

    self.addDevice('current',{
        probeType: 'condition',
        probeTitle: 'WeatherUndergoundCurrent',
        scaleTitle: scaleTemperature,
        title: self.langFile.current
    });

    self.addDevice('forecast',{
        probeType: 'forecast_range',
        probeTitle: 'WeatherUndergoundForecast',
        scaleTitle: scaleTemperature,
        title: self.langFile.forecast
    });

    if (self.config.forecastLowDevice === true) {
        self.addDevice('forecastLow',{
            probeType: 'forecast_low',
            icon: 'temperature',
            scaleTitle: scaleTemperature,
            title: self.langFile.forecastLow
        });
    }

    if (self.config.forecastHighDevice === true) {
        self.addDevice('forecastHigh',{
            probeType: 'forecast_high',
            icon: 'temperature',
            scaleTitle: scaleTemperature,
            title: self.langFile.forecastHigh
        });
    }

    if (self.config.humidityDevice === true) {
        self.addDevice('humidity',{
            probeType: 'humidity',
            icon: '/ZAutomation/api/v1/load/modulemedia/WeatherUnderground/humidity.png',
            scaleTitle: '%',
            title: self.langFile.humidity
        });
    }

    if (self.config.windDevice === true) {
        self.addDevice('wind',{
            probeType: 'wind',
            scaleTitle: config.unitSystem === "metric" ? 'km/h' : 'mph',
            title: self.langFile.wind
        });
    }

    if (self.config.uvDevice === true) {
        self.addDevice('uv',{
            probeType: 'ultraviolet',
            icon: 'ultraviolet',
            title: self.langFile.uv
        });
    }

    if (self.config.solarDevice === true) {
        self.addDevice('solar',{
            probeType: 'solar',
            scaleTitle: 'Watt/m²',
            icon: 'ultraviolet',
            title: self.langFile.solar
        });
    }

    if (self.config.barometerDevice === true) {
        self.addDevice('barometer',{
            probeType: 'barometer',
            scaleTitle: config.unitSystem === "metric" ? 'hPa' : 'inHg',
            icon: '/ZAutomation/api/v1/load/modulemedia/WeatherUnderground/barometer0.png',
            title: self.langFile.barometer
        });
    }

    var currentTime     = (new Date()).getTime();
    var currentLevel    = self.devices.current.get('metrics:level');
    var updateTime      = self.devices.current.get('metrics:timestamp');
    var intervalTime    = parseInt(self.config.interval,10) * 60 * 1000;

    self.timer = setInterval(function() {
        self.fetchWeather();
    }, intervalTime);

    setTimeout(function() {
        if (typeof(updateTime) === 'undefined') {
            self.fetchWeather();
        } else {
            console.log('[WeatherUnderground] Last update time '+updateTime);
            if ((updateTime + intervalTime / 3) < currentTime) {
                self.fetchWeather();
            }
        }
    },1000);
};

WeatherUnderground.prototype.stop = function () {
    var self = this;

    if (self.timer) {
        clearInterval(self.timer);
        self.timer = undefined;
    }

    if (typeof(self.devices) !== 'undefined') {
        _.each(self.devices,function(value, key) {
            self.controller.devices.remove(value.id);
        });
        self.devices = {};
    }

    if (typeof(self.update) !== 'undefined') {
        clearTimeout(self.update);
    }

    WeatherUnderground.super_.prototype.stop.call(this);
};

WeatherUnderground.prototype.addDevice = function(prefix,defaults) {
    var self = this;

    var probeTitle  = defaults.probeTitle || '';
    var scaleTitle  = defaults.scaleTitle || '';
    var probeType   = defaults.probeType || prefix;
    delete defaults.probeType;
    delete defaults.probeTitle;
    delete defaults.scaleTitle;

    var deviceParams = {
        overlay: {
            deviceType: "sensorMultilevel",
            probeType: probeType,
            metrics: {
                probeTitle: probeTitle,
                scaleTitle: scaleTitle
            }
        },
        defaults: {
            metrics: defaults
        },
        deviceId: "WeatherUnderground_"+prefix+"_" + this.id,
        moduleId: prefix+"_"+this.id,
        handler: function(command) {
            if (command === 'update') {
                if (typeof(self.update) !== 'undefined') {
                    clearTimeout(self.update);
                }
                self.update = setTimeout(_.bind(self.fetchWeather,self),10*1000);
            }
        }
    };

    self.devices[prefix] = self.controller.devices.create(deviceParams);
    return self.devices[prefix];
};

// ----------------------------------------------------------------------------
// --- Module methods
// ----------------------------------------------------------------------------

WeatherUnderground.prototype.fetchWeather = function (location, repeat) {
    var self = this;

    if (typeof(self.update) !== 'undefined') {
        clearTimeout(self.update);
    }

    repeat   = repeat || false;
    location = location || self.config.location;
    if (_.isArray(location)) {
        location = location[0];
    }
    var url  = "http://api.wunderground.com/api/"+self.config.apiKey+"/conditions/forecast/astronomy/q/"+location+".json";

    console.log("[WeatherUnderground] Fetch weather for "+location);

    http.request({
        url: url,
        async: true,
        success: function(response) {
            self.processResponse(response,repeat);
        },
        error: function(response) {
            console.error("[WeatherUnderground] Update error: "+response.statusText);
            console.logJS(response);
            self.controller.addNotification(
                "error",
                self.langFile.error_fetch,
                "module",
                self.constructor.name
            );
            if (location !== 'autoip') {
                self.fetchWeather('autoip',true);
            }
        }
    });
};

WeatherUnderground.prototype.processResponse = function(response,repeat) {
    console.log("[WeatherUnderground] Update");

    var self        = this;
    var current     = response.data.current_observation;
    var currentDate = new Date();
    var forecast    = response.data.forecast.simpleforecast.forecastday;
    var sun_phase   = response.data.sun_phase;
    var daynight    = 'day';
    if (typeof(sun_phase) !== 'undefined') {
        var sunrise     = response.data.sun_phase.sunrise;
        var sunset      = response.data.sun_phase.sunset;
        sunset.hour     = parseInt(sunset.hour,10);
        sunset.minute   = parseInt(sunset.minute,10);
        sunrise.hour    = parseInt(sunrise.hour,10);
        sunrise.minute  = parseInt(sunrise.minute,10);

        daynight = (
                currentDate.getHours() > sunrise.hour
                ||
                (
                    currentDate.getHours() === sunrise.hour
                    && currentDate.getMinutes() > sunrise.minute
                )
            )
            &&
            (
                currentDate.getHours() < sunset.hour
                ||
                (
                    currentDate.getHours() === sunset.hour
                    && currentDate.getMinutes() < sunset.minute
                )
            ) ? 'day':'night';
    }

    // Handle current state
    var maxDiff             = self.config.unitTemperature === "celsius" ? 7:13;
    var currentTemperature  = parseFloat(self.config.unitTemperature === "celsius" ? current.temp_c : current.temp_f);
    var currentHigh         = parseFloat(self.config.unitTemperature === "celsius" ? forecast[0].high.celsius : forecast[0].high.fahrenheit);
    var currentLow          = parseFloat(self.config.unitTemperature === "celsius" ? forecast[0].low.celsius : forecast[0].low.fahrenheit);
    var percipIntensity     = parseFloat(self.config.unitSystem === "metric" ? current.precip_1hr_metric : current.precip_1hr_in);
    var uv                  = parseInt(current.UV,10);
    var solarradiation      = parseInt(current.solarradiation,10);
    var temperatureList     = self.listSet(self.devices.current,"temperature_list",currentTemperature,3);
    var temperatureDiff     = _.last(temperatureList) - _.first(temperatureList);
    var changeTemperature   = 'unchanged';
    var humidity            = parseInt(current.relative_humidity,10);
    if (Math.abs(temperatureDiff) > 0.1) {
        if (temperatureDiff > 0) {
            changeTemperature = 'rise';
        } else {
            changeTemperature = 'fall';
        }
    }

    if (! repeat &&
        (humidity === -999 || humidity === -9999 || currentTemperature > currentHigh + maxDiff || currentTemperature < currentLow - maxDiff)) {
        console.error('[WeatherUnderground] Fallback to other weather station due to invalid data');
        self.fetchWeather(current.display_location.full,true);
        return;
    }

    self.devices.current.set("metrics:temperatureChange",changeTemperature);

    self.devices.current.set("metrics:conditiongroup",self.transformCondition(current.icon));
    self.devices.current.set("metrics:condition",current.icon);
    //self.devices.current.set("metrics:title",current.weather);
    self.devices.current.set("metrics:level",currentTemperature);
    self.devices.current.set("metrics:temperature",currentTemperature);
    self.devices.current.set("metrics:icon", "https://icons.wxug.com/i/c/k/"+(daynight === 'night' ? 'nt_':'')+current.icon+".gif");
    self.devices.current.set("metrics:feelslike", parseFloat(self.config.unitTemperature === "celsius" ? current.feelslike_c : current.feelslike_f));
    self.devices.current.set("metrics:weather",current.weather);
    self.devices.current.set("metrics:pop",forecast[0].pop);
    self.devices.current.set("metrics:high",currentHigh);
    self.devices.current.set("metrics:low",currentLow);
    self.devices.current.set("metrics:raw",current);
    self.devices.current.set("metrics:timestamp",currentDate.getTime());
    self.devices.current.set("metrics:percipintensity",percipIntensity);
    self.devices.current.set("metrics:uv",uv);
    self.devices.current.set("metrics:solarradiation",solarradiation);

    // Handle forecast
    var forecastHigh = parseFloat(self.config.unitTemperature === "celsius" ? forecast[1].high.celsius : forecast[1].high.fahrenheit);
    var forecastLow = parseFloat(self.config.unitTemperature === "celsius" ? forecast[1].low.celsius : forecast[1].low.fahrenheit);
    self.devices.forecast.set("metrics:conditiongroup",self.transformCondition(forecast[1].icon));
    self.devices.forecast.set("metrics:condition",forecast[1].icon);
    //self.devices.current.set("metrics:title",forecast[1].weather);
    self.devices.forecast.set("metrics:level", forecastLow + ' - ' + forecastHigh);
    self.devices.forecast.set("metrics:icon", "https://icons.wxug.com/i/c/k/"+forecast[1].icon+".gif");
    self.devices.forecast.set("metrics:pop",forecast[1].pop);
    self.devices.forecast.set("metrics:weather",forecast[1].conditions);
    self.devices.forecast.set("metrics:high",forecastHigh);
    self.devices.forecast.set("metrics:low",forecastLow);
    self.devices.forecast.set("metrics:raw",forecast);

    // Forecast low/high humidity
    if (self.config.forecastLowDevice === true) {
        self.devices.forecastLow.set("metrics:level", forecastLow);
    }
    if (self.config.forecastHighDevice === true) {
        self.devices.forecastHigh.set("metrics:level", forecastHigh);
    }

    // Handle humidity
    if (self.config.humidityDevice === true) {
        self.devices.humidity.set("metrics:level", humidity);
    }

    // Handle wind
    if (self.config.windDevice === true) {
        var windKph = parseInt(current.wind_kph,10);
        var windKphGust = parseInt(current.wind_gust_mph,10);
        var windMph = parseInt(current.wind_kph,10);
        var windMphGust = parseInt(current.wind_gust_mph,10);
        if (windKphGust > windKph) {
            windKph = windKph + (windKphGust - windKph) / 2;
            windMph = windKph + (windMphGust - windMph) / 2;
        }
        var currentWind = self.config.unitSystem === "metric" ? windKph : windMph;
        var beaufort = _.findIndex(self.windBeaufort,function(check) {
            return windKph < check;
        });
        var icon = _.findIndex(self.windIcons,function(check) {
            return beaufort < check;
        });
        self.devices.wind.set("metrics:icon", "/ZAutomation/api/v1/load/modulemedia/WeatherUnderground/wind"+icon+".png");
        self.devices.wind.set("metrics:dir", current.wind_dir);
        self.devices.wind.set("metrics:wind", parseFloat(self.config.unitSystem === "metric" ? current.wind_kph : current.wind_mph));
        self.devices.wind.set("metrics:windgust", parseFloat(self.config.unitSystem === "metric" ? current.wind_gust_kph : current.wind_gust_mph));
        self.devices.wind.set("metrics:winddregrees", parseFloat(current.wind_degrees));
        self.devices.wind.set("metrics:beaufort",beaufort);
        //var windList = self.listSet(self.devices.wind,"wind_list",currentWind,3);
        self.averageSet(self.devices.wind,currentWind,3);
    }

    // Handle UV
    if (self.config.uvDevice === true) {
        self.averageSet(self.devices.uv,uv,3);
    }

    // Handle solar intensity
    if (self.config.solarDevice === true) {
        self.averageSet(self.devices.solar,solarradiation,3);
    }

    // Handle barometer
    if (self.config.barometerDevice === true) {
        var pressure = parseFloat(self.config.unitSystem === "metric" ? current.pressure_mb : current.pressure_in);
        self.devices.barometer.set("metrics:icon", "/ZAutomation/api/v1/load/modulemedia/WeatherUnderground/barometer"+current.pressure_trend+".png");
        self.devices.barometer.set('metrics:level',pressure);
        self.devices.barometer.set('metrics:trend',current.pressure_trend);
    }
};

WeatherUnderground.prototype.transformCondition = function(condition) {
    if (_.contains(["chanceflurries", "chancesleet", "chancesnow", "flurries","sleet","snow"], condition)) {
        return 'snow';
    } else if (_.contains(["chancetstorms", "chancerain", "rain" ,"tstorms"], condition)) {
        return 'poor';
    } else if (_.contains(["cloudy", "mostlycloudy","fog"], condition)) {
        return 'neutral';
    } else if (_.contains(["clear", "hazy", "mostlysunny", "partlysunny", "partlycloudy"], condition)) {
        return 'fair';
    }

    return 'unknown';
};

WeatherUnderground.prototype.listSet = function(deviceObject,key,value,count) {
    var varKey = 'metrics:'+key;
    var list = deviceObject.get(varKey) || [];
    count = count || 3;
    list.unshift(value);
    while (list.length > count) {
        list.pop();
    }
    deviceObject.set(varKey,list);
    return list;
};

WeatherUnderground.prototype.averageSet = function(deviceObject,value,count) {
    var list = this.listSet(deviceObject,'list',value,count);
    var sum = _.reduce(list, function(i,j){ return i + j; }, 0);
    var avg = sum / list.length;
    deviceObject.set('metrics:current',avg);
    deviceObject.set('metrics:level',avg);
    return avg;
};



